### ISSUE TITLE
First test library manager

### ISSUE THREAD

**Issue by luni64** [luni64/VisualTeensy#1]
V0.5 : 
[VisualTeensy.zip](https://github.com/luni64/VisualTeensy/files/2412540/VisualTeensy.zip)

First test of shared and local library management. No GUI yet, need to test first if the makefile will work. 

- LIBS_SHARED_BASE -> Enter the path to a folder containing the shared libraries here
- LIBS_SHARED -> Add the names of the libraries you want to use. If the library contains a /src folder, add it to the library name as shown. 
- LIBS_LOCAL -> Add the names of libraries you copied to the lib folder in the project. If the library contains a /src folder, add it to the library name as shown. 

All those steps will be done by the GUI in a later version of course.

![image](https://user-images.githubusercontent.com/12611497/45970498-fbcbbb00-c036-11e8-9cea-f719127b609e.png)

**Comment by Defragster** [comment]
trying to run zip'd VisualTeensy.exe:
> doesn't recall my prior settings 
> Click folder icon to point to project folder and the App pauses and closes

**Comment by luni64** [comment]
Yep, that was a old version, the parallel make needs >v4.. I definitely need to clean up. Sorry for the confusion. [VisualTeensy.zip](https://github.com/luni64/VisualTeensy/files/2412574/VisualTeensy.zip)

**Comment by Defragster** [comment]
Builds and uploads as it was - gotta run now …

Will have to figure out the .mk edits to express the lib links for :: #include "debug_t3\debug_t3.h"

**Comment by luni64** [comment]
That should be easy. You just need to edit:
- LIBS_SHARED_BASE ->Path to the folder containing shared libraries. Typically that would be ...\documents\Arduino\library

- LIBS_SHARED :=  debug_t3.h  + any other library you need

**Comment by Defragster** [comment]
I did this:
LIBS_SHARED_BASE := T:\Programs\VisualTeensy\Source\libraries
LIBS_SHARED      := debug_t3.h

And got this - similar to before?
> Executing task: T:/Programs/VisualTeensy/make.exe uploadTy -j -Otarget <

make: *** No rule to make target 'src/debug_t3\debug_t3.h', needed by 'bin/src/main.o'.  Stop.
The terminal process terminated with exit code: 1

**Comment by luni64** [comment]
You need to enter the name of the library, not the name of the header...
LIBS_SHARED := debug_t3
without the .h

**Comment by Defragster** [comment]
Doh!

Okay next step:
> Executing task: T:/Programs/VisualTeensy/make.exe uploadTy -j -Otarget <

make: *** No rule to make target 'src/debug_t3\debug_t3.h', needed by 'bin/src/main.o'.  Stop.
make: *** Waiting for unfinished jobs....

**Comment by luni64** [comment]
It still tries to build a header?? Are you sure you changed to
LIBS_SHARED := debug_t3?
If yes please try a "clean" and then "build"

**Comment by luni64** [comment]
Can you upload your debug_t3 and the generated makefile?

**Comment by luni64** [comment]
..Or even better can you zip the complete folder (without the /bin subdirectory) and upload it here?

**Comment by Defragster** [comment]
I put both of these side by side in " T:\Programs\VisualTeensy\Source "

I did a CLEAN and it worked.

[libraries.zip](https://github.com/luni64/VisualTeensy/files/2412757/libraries.zip)
[DebugMin.zip](https://github.com/luni64/VisualTeensy/files/2412758/DebugMin.zip)

**Comment by luni64** [comment]
Just tried it. Works here as well....

But I had to fix the #inlcude statement in your main.cpp line 25 from #include "debug_t3\debug_t3.h" to the normal #include "debug_t3.h". That was proabably a leftover from the experiments from yesterday? 

![image](https://user-images.githubusercontent.com/12611497/45978109-af3eaa80-c04b-11e8-91a4-128203d39f96.png)

**Comment by luni64** [comment]
BTW: thanks for testing this. Looks like it is working in principle. Try to work on the GUI next. 

I'm not using the Arduino IDE much, so I don't know what Arduino users would expect.

- As far as I know installed Libs go into ...Documents/Arduino/library. So you'd usually use this as shared lib folder? 
- Do you think there is a need for having a second folder? 
- It wouldn't be too difficult to directly import libraries from GitHub. Does that make sense to you? Maybe combined with a search function? 
- Any other requirements for a cool library management tool?

**Comment by Defragster** [comment]
I did UNDO and took the folder name off and now it works?  I had just tried that and it failed from what I saw ...

Yes I put that on to make it work the OLD way.  

Since I do ZIP installs to allow multiple installs - then I point my sketchbook to "T:\tcode".  All my 1500 INO files are in there and there is a folder there too: "T:\tcode\libraries"

So yes, all 'personal' copies of library code go there and the IDE does its scan and locates them for inclusion as needed. Then there is only that one copy to update from github or other.

By default the IDE scans that SKETCHBOOK folder for anything needed, and will generally give it precedence over Teensy install libraries.

I don't use the default sketchbook in documents\Arduino as when I started on Teensy {IDE 1.6.0} that was seeming to have odd behavior - so I started all my local stuff in that tcode folder on a unique drive.

The IDE is great for opening throwaway sketches - copied from the forum and such. All coding assistance on the forum assumes code presented to put into the IDE to repro any issue.
